### PR TITLE
VSC Hardware Native Settings Grabbing

### PR DESCRIPTION
**Pull request by eyildirir**
# VSC Hardware Native Settings Grabbing

---

* Ticket: [SC-24779](https://app.shortcut.com/greenzie/story/24779/vsc-hardware-firmware-settings-grab)
* Developer(s): @eyildirir

## Description
This PR is one side of a pair of PRs that bring support for reading settings off of the Fort Robotic VSCs and pushing the data onto Gazano as a hardware version. The other part for the `robotic-lawnmower` can be found here:
https://github.com/Greenzie/robotic-lawnmower/pull/3492

This PR is the part that interfaces with the Fort Robotics VSCs to get the settings on board and provide them to other ROS nodes through a rosservice.

## Changes
* Added functionality for unlocking setup mode on the VSC.
* Added message templates and wrapper functions for querying settings.
* Added message handling for received settings.
* Updated list of recognized serial messages.
* Added new service for reporting VSC settings.

## Checklist:
- [x] Tested PR in `GZDEV13`
- [x] Tested PR in `GZDEV8`

### CODE REVIEW

File: include/hri_safe_remote_control_system/VehicleMessages.h
Comment (1hada) [inline include/hri_safe_remote_control_system/VehicleMessages.h:321]: just making sure the `SCMTargetGetMsgType` is supposed to be empty.

File: src/VscProcess.cpp
Comment (1hada) [inline src/VscProcess.cpp]: I understand those sleeps aren't very long. 
Wondering, Is there a way to wait for an `ACK-ACK` or `ACK-NACK` instead of sleeping?

File: src/VscProcess.cpp
Comment (1hada) [inline src/VscProcess.cpp]: Not sure on the rest of the member variables inside of the current repo, but in the mower repo we have been conforming to having member variables with underscores at the end.

`have_firmware_ && have_radio_power_level_ && have_firmware_`

Can you please update it to that since it makes it easier to read. Also the other ones added in the PR like 
`serial_` , `firmware_`, etc.

Update : 
Hmmm. noticing that the publishers and subscriber member variables are using the underscore convention. Might just need to hold off on it if there are too many member variables without that convention.

File: src/VscProcess.cpp
Comment (1hada) [inline src/VscProcess.cpp]:
```suggestion
  if (have_serial_ && have_firmware_ && have_radio_power_level_ && !srv_ready_)
```
Duplicate firmware check, and to add `have_serial_` to the checks.

**Review by 1hada — CHANGES_REQUESTED** [review]

File: include/hri_safe_remote_control_system/VehicleMessages.h
Comment (eyildirir) [reply include/hri_safe_remote_control_system/VehicleMessages.h:321]: It is. There's no payload to the message.

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: From my probing, I've not seen any acks back from the VSC. So currently, this is the only solution I can think of. But the waits may not be necessary between the `vsc_get_setting` queries. I'll do some testing on that.

File: CMakeLists.txt
Comment (kakaday22) [inline CMakeLists.txt]: could you alphabetize this please!

Comment (eyildirir) [reply CMakeLists.txt]: Done

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: Replaced duplicate `have_firmware` with `have_serial`, but kept  the non-underscore formatting to conform with the rest of the repo.

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: Removed sleep statements, and tested. It works asynchronously without them.

File: src/VehicleInterface.c
Comment (taylorwentzel) [inline src/VehicleInterface.c]: Is there something that you want to happen when this error occurs? Or will you print out some message in the future? Otherwise we can we remove the if statement and just send the message?

File: src/VehicleInterface.c
Comment (taylorwentzel) [inline src/VehicleInterface.c]: Same as above, will the commented out message be addressed in the future? Additionally, I see this happens several other times. So the same goes to these other instances.

File: src/VscProcess.cpp
Comment (taylorwentzel) [inline src/VscProcess.cpp]: Do we still need these commented out sleep statements? I'd go ahead and remove them if not.

File: src/VscProcess.cpp
Comment (taylorwentzel) [inline src/VscProcess.cpp]: Looks like these are member variables. I'd recommend appending `_` to the end of them for consistency.

**Review by taylorwentzel — COMMENTED** [review]
Left some comments.

**Review by taylorwentzel — DISMISSED** [review]
Requesting changes to get confirmation on whether you meant to merge into `noetic`.
Edit: my bad thought I was looking at robotic-lawnmower oof

**Comment by eyildirir** [comment]
> Requesting changes to get confirmation on whether you meant to merge into `noetic`.

Well, we don't have a `noetic-devel` for this submodule, so yes, probably. There's nowhere else for it to go...

File: src/VehicleInterface.c
Comment (eyildirir) [reply src/VehicleInterface.c]: That error handling was inherited from how the other previously existing handler functions handled the errors. I'm going to defer to the judgement of the previous developer here, so that we have an easy way of debugging if something doesn't work in the future.

File: src/VehicleInterface.c
Comment (eyildirir) [reply src/VehicleInterface.c]: Same as above, will defer to the previous developer's judgement on this.

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: Good catch, will remove.

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: The rest of the repo, including every other file doesn't follow this convention, so it would be out of the scope of this PR to change every other preexisting variable along with these as well. I believe we should opt to keep them as is to conform to the submodule's own formatting rules for now.

File: src/VscProcess.cpp
Comment (eyildirir) [reply src/VscProcess.cpp]: Changing all of the previously existing formatting to conform with the `robotic-lawnmower` repo's formatting would probably be out of the scope of this PR, and we should probably have another PR for it, if necessary. I'd rather opt to keep the formatting uniform within the submodule for now.

File: src/VscProcess.cpp
Comment (1hada) [reply src/VscProcess.cpp]: @taylorwentzel I recommended the same thing but, I agree with Eren that if it doesn't conform with the style previously inheritted from HRI then it might be fine as is.

File: src/VehicleInterface.c
Comment (1hada) [reply src/VehicleInterface.c]: I would prefer if it actually printed.

File: srv/GetVscSettings.srv
Comment (1hada) [inline srv/GetVscSettings.srv]:
```suggestion
int16 radio_power_level_db
```
I like that you added the comment. In thinking about it more and seeing how it is used int he other PR, can you please add the `_db` where ever that `radio_power_level` is used.

**Review by 1hada — CHANGES_REQUESTED** [review]
2 requests for change. 
1. un-comment the prints so we can see if something goes wrong in the requests.
2. documentation via variable names (also keep the comment since seeing the "in dB" is helpful to confirm the units.)

File: src/VehicleInterface.c
Comment (eyildirir) [reply src/VehicleInterface.c]: Did some testing with printing, and what I was afraid of came true... I tested some VSC disconnections, and the message spams at a 50Hz rate, clogging up the tail log. I think it might just be better to remove it altogether like @taylorwentzel suggested. It's not like we're going to get any real use our of it...

File: src/VehicleInterface.c
Comment (eyildirir) [reply src/VehicleInterface.c]: Removed, as discussed above.

File: srv/GetVscSettings.srv
Comment (eyildirir) [reply srv/GetVscSettings.srv]: Decided to use `radio_power_db` for shortness.

**Review by 1hada — APPROVED** [review]

**Review by taylorwentzel — APPROVED** [review]
Nice job removing unnecessary stuff. Looks good!

**Pull request merged**
